src/protocols/http/hpr_http_roaming_worker.erl

```diff
@@ -0,0 +1,258 @@
+        send_data_timer = DataTimeout
+    } = State0
+) ->
+    lager:debug("packet delivery after data sent [send_data_timer: ~w]", [DataTimeout]),
+    {ok, ShutdownTimerRef1} = maybe_schedule_shutdown(ShutdownTimerRef0),
+    {noreply, State0#state{shutdown_timer_ref = ShutdownTimerRef1}};
+handle_cast(_Msg, State) ->
+    lager:warning("rcvd unknown cast msg: ~p", [_Msg]),
+    {noreply, State}.
+
+handle_info(?SEND_DATA, #state{} = State) ->
+    ok = send_data(State),
+    {ok, ShutdownTimerRef} = maybe_schedule_shutdown(undefined),
+    {noreply, State#state{should_shutdown = true, shutdown_timer_ref = ShutdownTimerRef}};
+handle_info(?SHUTDOWN, #state{} = State) ->
+    {stop, normal, State};
+handle_info(_Msg, State) ->
+    lager:warning("rcvd unknown info msg: ~p, ~p", [_Msg, State]),
+    {noreply, State}.
+
+terminate(_Reason, #state{}) ->
+    lager:info("going down ~p", [_Reason]),
+    ok.
+
+%% ------------------------------------------------------------------
+%% Internal Function Definitions
+%% ------------------------------------------------------------------
+
+-spec maybe_schedule_send_data(integer(), undefined | reference()) -> {ok, reference()}.
+maybe_schedule_send_data(Timeout, undefined) ->
+    {ok, erlang:send_after(Timeout, self(), ?SEND_DATA)};
+maybe_schedule_send_data(_, Ref) ->
+    {ok, Ref}.
+
+-spec maybe_schedule_shutdown(undefined | reference()) -> {ok, reference()}.
+maybe_schedule_shutdown(undefined) ->
+    {ok, erlang:send_after(?SHUTDOWN_TIMEOUT, self(), ?SHUTDOWN)};
+maybe_schedule_shutdown(CurrTimer) ->
+    _ = (catch erlang:cancel_timer(CurrTimer)),
+    {ok, erlang:send_after(?SHUTDOWN_TIMEOUT, self(), ?SHUTDOWN)}.
+
+-spec next_transaction_id() -> integer().
+next_transaction_id() ->
+    rand:uniform(16#7FFFFFFF).
+
+-spec do_handle_packet(
+    PacketUp :: hpr_packet_up:packet(),
+    GatewayTime :: hpr_http_roaming:gateway_time(),
+    GatewayStream :: hpr_router_stream_manager:gateway_stream(),
+    State :: #state{}
+) -> {ok, #state{}}.
+do_handle_packet(
+    PacketUp, GatewayTime, GatewayStream, #state{packets = Packets} = State
+) ->
+    State1 = State#state{
+        packets = [
+            hpr_http_roaming:new_packet(PacketUp, GatewayTime, GatewayStream) | Packets
+        ]
+    },
+    {ok, State1}.
+
+-spec send_data(#state{}) -> ok.
+send_data(
+    #state{
+        net_id = NetID,
+        address = Address,
+        packets = Packets,
+        transaction_id = TransactionID,
+        flow_type = FlowType,
+        auth_header = Auth,
+        protocol_version = ProtocolVersion,
+        send_data_timer = DedupWindow
+    }
+) ->
+    Data = hpr_http_roaming:make_uplink_payload(
+        NetID,
+        Packets,
+        TransactionID,
+        ProtocolVersion,
+        DedupWindow,
+        Address,
+        FlowType
+    ),
+    RoundedFloats = semtech_udp:round_to_fourth_decimal_all_float_values(Data),
+    Data1 = jsx:encode(RoundedFloats),
```

Comment:
I don't know that this solves the same problem as the original.

The float encoding that was used inside `jsx` output the string that would be inserted directly into the serialized binary.

`semtech_udp:round_to_fourth_decimal_all_float_values/1` speeds through the map looking for floats and truncating them to 4 decimal places, but still leaves them as floats. When they get to `jsx:encode/1` they have the same chance of being improperly serialized as if we did nothing.

And another case of crossing the protocol boundary, This isn't udp.

Reply:
Moved truncation into hpr_packet_up:frequency_mhz/1.